### PR TITLE
materialized: reimplement telemetry reporting in terms of SQL

### LINKED ISSUE #7137: telemetry not collected for Kafka or PG sources
(to fill in)

### LINKED ISSUE #7181: Telemetry counts are reset to 0 after restart
### What version of Materialize are you using?

```
materialized v0.8.1-dev (da38e0c97)
```

### How did you install Materialize?

* [x] Built from source

### What was the issue?

Telemetry counts are reset after restart.

After creating a file source the telemetry report includes `"file":{"count": 1}`.

```json
 {"version":"0.8.1-dev","status":{"session_id":"696fd58a-a3a4-4789-87ee-1111bdbdb383","uptime_seconds":90,"num_workers":4.0,"sources":{"avroOcf":{"count":0},
"file":{"count":1},"kinesis":{"count":0},"postgres":{"count":0},"s3":{"count":0},"table":{"count":0}},"views":{"count":0},"sinks":{"tail":{"count":0},"kafka":{"count":0},"avroOcf":{"count":0}}}}
```

After `materialized` restarts, the count is reset to 0.

```json
{"version":"0.8.1-dev","status":{"session_id":"463d1e1c-9a48-42da-9d0a-8887a6012258","uptime_seconds":0,"num_workers":4.0,"sources":{"avroOcf":{"count":0},
"file":{"count":0},"kinesis":{"count":0},"postgres":{"count":0},"s3":{"count":0},"table":{"count":0}},"views":{"count":0},"sinks":{"tail":{"count":0},"kafka":{"count":0},"avroOcf":{"count":0}}}}
```

This could be by design, but our internal dashboard seems to assume that counts are not reset. cc @awang 

---

This has been discovered with @aljoscha during dogfooding day.

### PR DESCRIPTION
**Pull request by benesch**
I cc'd everyone who's ever touched telemetry on this review. Probably overkill, but didn't want to leave anyone out.

----

Rework telemetry reporting to use the SQL system tables instead of
Prometheus metrics. This avoids double-bookkeeping information for
information that can be derived from the system tables, like the number
of sources and sinks in the system and their types. Forcing all
telemetry information to come from the system catalog is also a nice
pressure to ensure we surface information to our users; if we want to
collect some metadata for telemetry, there's a good chance our users
want to be able to collect that metadata too.

I rewrote the version check loop to report new versions not just at
startup but also if a new version is discovered during telemetry
reporting at any point in the server's lifetime. As a result it's
probably easiest just to review the changes in telemetry.rs as new code.

Perhaps most importantly, collecting telemetry this way makes it much
harder to wind up with bugs like those below.

Fix #7137.
Fix #7181.

### CODE REVIEW

File: src/materialized/src/telemetry.rs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+                "a new version of materialized is available: {}",
+                latest_version
+            );
+            reported_version = latest_version;
+        }
+    }
+}
+
+/// The query used to gather telemetry data.
+const TELEMETRY_QUERY: &str = "SELECT jsonb_build_object(
```
Comment (mjibson) [inline src/materialized/src/telemetry.rs:72]: This is so good. More interfaces in SQL!

**Review by mjibson — APPROVED** [review]

**Comment by umanwizard** [comment]
nice

File: src/materialized/src/telemetry.rs (hunk 1 above)
Comment (uce) [reply src/materialized/src/telemetry.rs:76]: This is great! 👍🏼

**Review by uce — APPROVED** [review]
LGTM. This is really elegant 😊

Note that we now report all system tables whereas previously we only reported user-created tables under `tables` (modulo bugs). I think this is good, but wanted to flag it in case it was unintended.

**Comment by benesch** [comment]
> Note that we now report all system tables whereas previously we only reported user-created tables under `tables` (modulo bugs). I think this is good, but wanted to flag it in case it was unintended.

Whoops! That wasn't intended. I've put it back to the old behavior so we can consider if we want to make a behavior change separately from this PR.

**Comment by benesch** [comment]
Thanks for the reviews, everyone!

**Pull request merged**